pingrid/impl.py

```diff
@@ -96,24 +87,40 @@ class ColorScale:
     
     def __init__(self, colors, scale=None):
         self.colors = colors
-        self.scale = scale
-    
-    def cv2cm(self):
-        if self.scale is None:
-           cs_val = np.arange(len(colors))
+        if scale is None:
+            self.scale = list(np.arange(len(colors)))
         else:
-           if len(self.colors) == len(self.scale):
-               cs_val = np.array(self.scale)
+           if len(colors) == len(scale):
+               self.scale = scale
            else:
-               raise Exception("if provided, scale must be same length as colors")
-        cv2cm_i = (255*(cs_val-cs_val[0])/(cs_val[-1]-cs_val[0])).astype(int)
-        cv2cm_i = cv2cm_i + np.append([0], np.where(np.diff(cv2cm_i) == 0, 1, 0))
-        cv2cm_rgba = np.array(self.colors)
-        cv2cm_bgra = cv2cm_rgba[:,[2, 1, 0, 3]]
-        cv2cm = np.full((256,4), np.nan)
-        for bgra in range(4):
-            cv2cm[:, bgra] = np.interp(np.arange(256), cv2cm_i, cv2cm_bgra[:, bgra])
-        return cv2cm.astype(int)
+               raise Exception(
+                   "if provided, scale must be same length as colors"
```

Comment:
Here is some ASCII art explaining why I think you need n-1 thresholds to go with n colors. If you have n thresholds, I don't know where to put the last one in this diagram.
```
       t1   t2   t3
<-- c1 | c2 | c3 | c4 -->
```

Reply:
Put in words, the thresholds represent points on the continuum, while the colors correspond to intervals. It takes n - 1 points to partition the continuum into n intervals.

Reply:
That's not how things are done right now, or intended to be done. Color ci is set at Ii that corresponds to ti and then interpolated for the rest of the array. If ti = tj, Ij = Ii +1

Reply:
I was (wrongly) thinking of bars of constant color rather than interpolation, but it seems like if you add interpolation to the mix, and you allow for color discontinuities at the boundaries, then you need even fewer thresholds for a given number of colors:

```
       t1         t2
<-- c1 | c2 -- c3 | c4 -->
```

As I mentioned yesterday, the colorscale code was originally written assuming a bounded range,  e.g. probabilities in [0-1], and then inappropriately used to display unbounded variables, e.g. precipitation amounts. Then I made a quick fix to make the precip maps not look like random noise when precip exceeded the expected maximum, but I was in "quick fix" mode at the time, and I may well have done it wrong or in a way that wasn't fully general. Let's try to get it right this time.

This kind of detail is why I thought it might be useful to do a survey of how each library treats colormaps, and see if we can pick a representation that can be mapped from ingrid and to any of plotly, opencv, and matplotlib.

Reply:
Oh, I see, the colormap I described above can be represented as follows in the current representation:

```
    t1   t1    t2   t2
<-- c1 | c2 -- c3 | c4 -->
```

So in terms of expressiveness it's equivalent, if a bit less efficient. But now I wonder if we're actually translating it correctly into opencv and plotly, including for these complicated cases like unbounded variables and discontinuities at boundaries. Do you know if we are?

Reply:
I don't know. I took inspiration from matplotlib representations that do allow this:

-- a list of anchors or nodes with colors defined on each side of the anchor/node, e.g.:
_autumn_data = {'red': ((0., 1.0, 1.0), (1.0, 1.0, 1.0)),
'green': ((0., 0., 0.), (1.0, 1.0, 1.0)),
'blue': ((0., 0., 0.), (1.0, 0., 0.))}

But also allows this:

-- but matplotlib doc indicated the following to create a colorscale through anchor points:
colors = ["darkorange", "gold", "lawngreen", "lightseagreen"]
nodes = [0.0, 0.4, 0.8, 1.0]
cmap2 = LinearSegmentedColormap.from_list("mycmap", list(zip(nodes, colors)))

and from the output of pingrid.parse_colormap.

I don't know if the threshold/interval is the way to think. If c1=c2, you don't want the color to take 2 consecutive spots in the array, don't you? On the other hand, since we are limited to 255 colors and integers from 0 to 255, and then scaling to whatever possible range of numbers, it would be more correct to rely on a threshold with colors on each side.

My understanding of things is superficial and approximative at this point an I only try to reproduce what I could see in pingrid, but starting from a different starting point. So I will have to look into more details into what plotly, opencv and matplotlib do and expect.

Reply:
and from what I can tell, pingrid.apply_colormap is rescaling data against so that scale_min becomes 0 and scale_max becomes 255 and the clips at 0 and 255 so there is no data point below/above the min/max.

Reply:
Yeah, I added the clipping. I don't know if that's the right way to deal with out-of-range values.

Reply:
> I don't know if the threshold/interval is the way to think. If c1=c2, you don't want the color to take 2 consecutive spots in the array, don't you? 

That's what I originally took issue with, but now I see that eliminating that redundancy would make it less expressive, so I'm ok with it.

What other way to think about it is there, other than threshold/interval?

Reply:
Ok. let's make a pause and rewind a bit.

We have data that goes from r0 to rn that are real numbers, ro <= rn
We want to put such data on a scale that goes from smin to smax, other real numbers, and smin <= smax. However, the relative position of r0 and rn with respect to smin and smax could be whatever (to generalize).

We are allowed 256 colors represented as an array of RGBA, or hex-stuff, or some other stuff I guess.

First color (0) should be assigned to data values < smin; last color (255) assigned to data values > smax. That leaves 254 colors from smin to smax, included.

So we are rescaling data from [smin, smax] to [1, 254].

Then we construct our array of 254 colors with some constraints and... I have to go take the shuttle...

Reply:
Then I guess that's where I am not sure what is supposed to happen. We want data value x assigned a specific color, thus assigned an integer between 1 and 254. Do we round the scaled x to the closest integer? Do we make small bins and assigned scaled x to one?

Reply:
Also we want to define colorscales inputting data values as thresholds (then be able to reuse an existing colorscale rescaling it for other data) so there is an issue of whether thresholds are inclusive or not, especially when making bands and 2 colors are assigned same value, which one should be of the color? And I suspect there might be 2 cases to consider: whether the data can take more or less than 256 values between smin and smax included. It's probably more subtle than that.

Then there is eventually the case of categorical data.

Reply:
I've spent several hours writing this comment, and come to understand things better in the process. The bottom line is that I think you're on the right track, it just took me some time to catch up with you.

Let's note that r0...rn may not be known at the moment that the colormap is defined. (I don't think that invalidates anything you said.)

> Do we round the scaled x to the closest integer? Do we make small bins and assigned scaled x to one?

Bins. Value x maps to colors[y] where y = floor((x-min) * 256 / (max-min)), assuming min <= x <= max. This is implemented in pingrid.apply_colormap. opencv requires its colormap to be in this 256-entry form.

I'm not sure we need to assume that n=256. That's a common value, and the only value allowed in opencv, but matplotlib seems to support other values. I like the anchor points representation because it doesn't constrain the number of values _a priori_; quantization can be delayed until we call into opencv, and if one day we replace opencv with something that supports higher color resolutions, we will be able to take advantage of it without changing most of our code.

Regarding out-of-range values, I surveyed the docs for the three packages.
- matplotlib uses n (256 by default) colors for in-range values, and an additional three for below min, above max, and masked. It doesn't reserve colors 0 and 255 for out-of-range values as you suggested.  By default it uses color 0 for below min and color 255 for above max (in addition to using those same colors for the first and last in-range intervals), but you can override that.
- Plotly maps below-min values to color 0 and above-max to color 255 with no option to override as far as I can see.
- Opencv doesn't seem to deal with out-of-range values at all. You're expected to normalize and clip your data before passing it to opencv for rendering, and if you don't, colors end up wrapping around, which is nonsensical. To avoid strange and library-dependent behavior, let's always clip data to min and max before passing it to any third-party library.

I see that plotly's colorscale representation is the same as the one you're using (n values, n colors), except it seems to assume that the values will all be between 0 and 1. I tried using values other than 0 and 1 for the first and last threshold, and plotly didn't complain, but I can't make sense of the result. Your approach of using arbitrary values and mapping them into [0, 1] or [0, 255] behind the scenes seems better.

> there is an issue of whether thresholds are inclusive or not, especially when making bands and 2 colors are assigned same value, which one should be of the color? 

The intervals should be considered open on one end and closed on the other, so that they don't overlap. Closed at the low end and open at the high end seems the most natural to me, but you could check what plotly and matplotlib do.

> And I suspect there might be 2 cases to consider: whether the data can take more or less than 256 values between smin and smax included. It's probably more subtle than that.

I don't understand what you're saying there. Could you elaborate or give an example?

Big picture:

Let's distinguish two things that we may have been conflating so far: the colormap representation our library uses internally, which can be translated to and from the representations used by other software; and the representation in which users of our library specify their colormaps. In the current implementation the two are the same, but they don't have to be. Plotly and matplotlib each offer multiple ways of specifying a colormap that are convenient in different cases, and we could do that too.

The anchor points representation you're using seems like a good internal representation. I objected to it at first because it seemed redundant, but now I understand that it gives greater expressiveness than the alternative I was suggesting, so I no longer object. (Specifically, it allows for interpolated color scales with discontinuities.)

Since the anchor points representation is redundant in simple cases, it would be nice to have the option to specify colormaps in a simpler format, as both matplotlib and plotly allow. But that can come later if we want.

I'm not sure this comment is going to help. Maybe we should talk.

Reply:
it's very helpful. I can see that the conversation can branch out on different sub-topics from here so maybe yes indeed it will be more efficient to talk. Maybe Tuesday January 10th? assuming we're both in the office and can use the blackboard? That is only 3 working-days from today, for me, after all!

Reply:
Shall we assume an in-person chat in the office next Tuesday (17th)?